pkg/agent/controller/supportbundlecollection/support_bundle_controller.go

```diff
@@ -0,0 +1,357 @@
+		return err
+	}
+	if err = agentDumper.DumpOVSPorts(basedir); err != nil {
+		return err
+	}
+
+	outputFile, err := afero.TempFile(defaultFS, "", "bundle_*.tar.gz")
+	if err != nil {
+		return fmt.Errorf("error when creating temp file: %w", err)
+	}
+	defer func() {
+		if err := outputFile.Close(); err != nil {
+			klog.Errorf("Error occurs when closing output tar file: %v", err)
+		}
+		if err := defaultFS.Remove(outputFile.Name()); err != nil {
+			klog.Errorf("Error occurs when removing output tar file %s: %v", outputFile.Name(), err)
+		}
+
+	}()
+	if _, err := compress.PackDir(defaultFS, basedir, outputFile); err != nil {
+		return fmt.Errorf("error when packaging support bundle: %w", err)
+	}
+
+	return c.uploadSupportBundle(supportBundle, outputFile)
+}
+
+func (c *SupportBundleController) uploadSupportBundle(supportBundle *cpv1b2.SupportBundleCollection, outputFile afero.File) error {
+	if _, err := outputFile.Seek(0, 0); err != nil {
+		return fmt.Errorf("support bundle upload to file server failed during setting offset: %v", err)
+
+	}
+	triesLeft := UPLOAD_TO_FILESERVER_TRIES
+	var uploadErr error
+	for triesLeft > 0 {
+		if uploadErr = c.uploadToFileServer(supportBundle.Name, &supportBundle.FileServer, &supportBundle.Authentication, outputFile); uploadErr != nil {
+			triesLeft -= 1
+			klog.Warningf("Failed to upload support bundle by SFTP: %v, triesLeft: %d", uploadErr, triesLeft)
+			time.Sleep(UPLOAD_TO_FILESERVER_RETRY_DELAY)
+		} else {
+			return nil
+		}
+	}
+	return fmt.Errorf("failed to upload support bundle by SFTP: %w", uploadErr)
+}
+
+func (c *SupportBundleController) uploadToFileServer(bundleName string, fileServer *cpv1b2.BundleFileServer, serverAuth *cpv1b2.BundleServerAuthConfiguration, tarGzFile io.Reader) error {
+	// fileServer.URL should be like: 10.92.23.154:22/path or sftp://10.92.23.154:22/path
+	var url string
+	if strings.Contains(fileServer.URL, "://") {
+		arr := strings.Split(fileServer.URL, "://")
+		if strings.ToLower(arr[0]) != "sftp" {
+			return fmt.Errorf("only sftp is supported for now, protocol %s is not supported", strings.ToLower(arr[0]))
+		}
+		url = fileServer.URL[7:]
+	} else {
+		url = fileServer.URL
+	}
+	strArr := strings.Split(url, "/")
+	address := strArr[0]
+	targetPath := url[len(address):]
+	path := path.Join(targetPath, c.nodeName+"_"+bundleName+".tar.gz")
+	cfg := &ssh.ClientConfig{
+		User: serverAuth.BasicAuthentication.Username,
+		Auth: []ssh.AuthMethod{ssh.Password(serverAuth.BasicAuthentication.Password)},
+		// Skip host key check here and users can specify their own checks if needed
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		Timeout:         time.Second,
+	}
+	uploader, err := getUploader("sftp")
+	if err != nil {
+		return err
+	}
+	return uploader.upload(address, path, cfg, tarGzFile)
+}
+
+var getUploader = getUploaderByProtocol
+
+func getUploaderByProtocol(protocol string) (uploader, error) {
+	if protocol == "sftp" {
+		return &sftpUploader{}, nil
```

Comment:
Think about using singleton instance for uploader. No need to create a new object each time calling `getUploaderByProtocol`